Error if DB downgrades are required

johnswanson opened the pull request:
Before running migrations, check the major versions of:

- the latest database migration that was actually applied (e.g. if v49 ran migrations, this would be 49)

- the latest database migration that liquibase is aware of (e.g. if we're currently running v48, this would be 48)

If the database has executed migrations that we don't know about, throw an exception and exit with a helpful error message telling the user they need to run `migrate down` from the newer version before they can downgrade.

Fixes https://github.com/metabase/metabase/issues/27906

johnswanson committed:
Error if DB downgrades are required

Before running migrations, check the major versions of:

- the latest database migration that was actually applied (e.g. if v49
ran migrations, this would be 49)

- the latest database migration that liquibase is aware of (e.g. if
we're currently running v48, this would be 48)

If the database has executed migrations that we don't know about, throw
an exception and exit with a helpful error message telling the user they
need to run `migrate down` from the newer version before they can
downgrade.

Fixes #27906

noahmoss reviewed (approved):
Functionality lgtm. 

My one question is about the error message — I wonder if we should link to the [relevant page in the docs](https://www.metabase.com/docs/latest/installation-and-operation/upgrading-metabase#rolling-back-an-upgrade).

Might be a question for @jeff-bruemmer and/or the success team. I think we have a couple other places where we have hard-coded docs links in the codebase, but not sure if that's encouraged.

jeff-bruemmer commented:
> My one question is about the error message — I wonder if we should link to the [relevant page in the docs](https://www.metabase.com/docs/latest/installation-and-operation/upgrading-metabase#rolling-back-an-upgrade).

It's okay (and can be better) to link to documentation. One consideration, however, is embedding/white-labeling, but this particular case is admin territory, where white-labeling doesn't apply.

johnswanson committed:
CR: Add link to error message on downgrade detect

(and a newline removal)

johnswanson commented on src/metabase/db/setup.clj:
```diff
@@ -53,6 +53,7 @@
                    "\n"))
     (throw (Exception. (trs "Database requires manual upgrade.")))))
 
+
```
Apparently this is a habit of mine 😂

johnswanson committed:
oops: use `log/error` for the logged... error

I put the link in the log emitted just before the exception is thrown,
but then realized on a second look that if someone has the log level set
to something higher than `INFO`, they'll just get
`ERROR: Downgrade detected`, which is unclear.

johnswanson commented:
> My one question is about the error message — I wonder if we should link to the [relevant page in the docs](https://www.metabase.com/docs/latest/installation-and-operation/upgrading-metabase#rolling-back-an-upgrade).

@noahmoss does this look good? Specifically, I added the link to the error message emitted just *before* the error itself is thrown - just wanted to check that that makes sense to you too.

noahmoss commented:
@johnswanson Looks good!
